File: 78.子集.cpp
/*
 * @lc app=leetcode.cn id=78 lang=cpp
 *
 * [78] 子集
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> current;
        foo(nums, 0, current, res);
        return res;
    }
    void foo(const vector<int>& nums, int start_idx, 
            const vector<int>& current,
            vector<vector<int>>& res) {
        res.push_back(current);
        if (start_idx >= nums.size()) {
            return;
        }
        for (int i = start_idx; i < nums.size(); ++i) {
            auto next = current;
            next.push_back(nums[i]);
            foo(nums, i + 1, next, res);
        }
    }
};
// @lc code=end

File: 78.子集.py
#
# @lc app=leetcode.cn id=78 lang=python
#
# [78] 子集
#

# @lc code=start
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def foo(target, cur, res):
            res.append(cur)
            if not target:
                return
            for i in target:
                tmp = target[:]
                tmp.remove(i)
                if not cur or i > cur[-1]:
                    foo(tmp, cur + [i], res)
        
        res = []
        foo(nums, [], res)
        return res
# @lc code=end

